### PR TITLE
Handle socket timeout exception

### PR DESCRIPTION
**Pull request by anjor**
Retries on socket timeout exception.

### CODE REVIEW

File: atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxy.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        InvocationTargetException exception = (InvocationTargetException) throwable;
+        Throwable cause = exception.getCause();
+        if (!isCausedBySocketTimeoutException(cause)) {
+            return ResultOrThrowable.failure(cause);
+        }
+        return ResultOrThrowable.success(null);
+    }
+
+    @VisibleForTesting
+    static boolean isCausedBySocketTimeoutException(Throwable throwable) {
```
Comment (jeremyk-91) [inline atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxy.java:86]: we should add simple tests

**Review by jeremyk-91 — COMMENTED** [review]
Looks good! I think we should add some tests, but otherwise we should be close

File: atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxy.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            return ResultOrThrowable.failure(th);
+        }
+    }
+
+    private ResultOrThrowable isRetriable(Throwable throwable) {
+        if (!(throwable instanceof InvocationTargetException)) {
+            return ResultOrThrowable.failure(throwable);
+        }
+        InvocationTargetException exception = (InvocationTargetException) throwable;
+        Throwable cause = exception.getCause();
```
Comment (Jolyon-S) [inline atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxy.java:78]: Is this necessary? Can you not just get a cause from a generic throwable?

File: atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxy.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        if (!isCausedBySocketTimeoutException(cause)) {
+            return ResultOrThrowable.failure(cause);
+        }
+        return ResultOrThrowable.success(null);
```
Comment (Jolyon-S) [inline atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxy.java:82]: Flip the if, and I'd add the other part in the else block.

```suggestion
        if (isCausedBySocketTimeoutException(cause)) {
            return ResultOrThrowable.success(null);
        } else {
            return ResultOrThrowable.failure(cause);
        }
```

File: atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxy.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        return (U) Proxy.newProxyInstance(interfaceClass.getClassLoader(), new Class<?>[] {interfaceClass}, proxy);
+    }
+
+    @Override
+    protected Object handleInvocation(Object proxy, Method method, Object[] args) throws Throwable {
+        ResultOrThrowable attempt = singleInvocation(method, args);
+        while ((numRetries < MAX_NUM_RETRIES) && !attempt.isSuccessful()) {
+            numRetries++;
+            Throwable cause = attempt.throwable().get();
+            ResultOrThrowable retriableExceptionCheck = isRetriable(cause);
```
Comment (Jolyon-S) [inline atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxy.java:53]: I'm really not a fan of how you are using this here - it's never actually a result, but rather it is either retriable, or it is not and there is an exception. You could just return `Optional<Throwable>` instead for the same effect.

File: atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxy.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+        return (U) Proxy.newProxyInstance(interfaceClass.getClassLoader(), new Class<?>[] {interfaceClass}, proxy);
+    }
+
+    @Override
+    protected Object handleInvocation(Object proxy, Method method, Object[] args) throws Throwable {
+        ResultOrThrowable attempt = singleInvocation(method, args);
+        while ((numRetries < MAX_NUM_RETRIES) && !attempt.isSuccessful()) {
+            numRetries++;
+            Throwable cause = attempt.throwable().get();
```
Comment (Jolyon-S) [inline atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxy.java:52]: There is no form of backoff between retries; do we want that? I'm a bit worried that we're just going to still fail 6 times in rapid succession.

File: atlasdb-conjure/src/test/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxyTest.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+package com.palantir.atlasdb.http.v2;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.net.SocketException;
+import java.net.SocketTimeoutException;
+import org.junit.Test;
+
+public class RetryOnSocketTimeoutExceptionProxyTest {
```
Comment (Jolyon-S) [inline atlasdb-conjure/src/test/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxyTest.java:25]: Can you add some tests that mock the supplier to fail say 3 times, then succeed, and verify that it passes? For example:

```java
Supplier<Long> supplier = mock(Supplier.class);
when(supplier.get())
    .thenThrow(new RuntimeException(new SocketTimeoutException()))
    .thenThrow(...)
    .thenThrow(...)
    .thenReturn(11L);

```

Comment (Jolyon-S) [reply atlasdb-conjure/src/test/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxyTest.java:25]: (And some other tests of this form for no failures, only failures, failures where exceptions don't have the right root cause, etc.)

**Review by Jolyon-S — COMMENTED** [review]
Wiring looks right, I have a couple of concerns and a few test suggestions.

File: atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxy.java (hunk 4 above)
Comment (Jolyon-S) [reply atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxy.java:53]: ugh, FFP does the same.

File: atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxy.java (hunk 5 above)
Comment (Jolyon-S) [reply atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxy.java:52]: actually no, FFP.

File: atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxy.java (hunk 2 above)
Comment (Jolyon-S) [reply atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxy.java:78]: FFP code...

File: atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxy.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+import java.util.function.Supplier;
+
+/**
+ * This proxy exists to retry in case of a SocketTimeoutException (something that isn't retried natively by Dialogue).
+ */
+public final class RetryOnSocketTimeoutExceptionProxy<T> extends AbstractInvocationHandler {
+    private static final int MAX_NUM_RETRIES = 5;
+    private final Supplier<T> delegate;
+
+    private static int numRetries = 0;
```
Comment (Jolyon-S) [inline atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxy.java:35]: ooooh this looks weird

does FFP use a `static` here?!

File: atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxy.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        while ((numRetries < MAX_NUM_RETRIES) && !attempt.isSuccessful()) {
+            numRetries++;
+            Throwable cause = attempt.throwable().get();
+            ResultOrThrowable retriableExceptionCheck = isRetriable(cause);
+            if (!retriableExceptionCheck.isSuccessful()) {
+                throw retriableExceptionCheck.throwable().get();
+            }
+            attempt = singleInvocation(method, args);
+        }
+        numRetries = 0;
```
Comment (Jolyon-S) [inline atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxy.java:59]: Probably makes more sense to reset this to 0 at the beginning of the loop?

Comment (Jolyon-S) [reply atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxy.java:59]: Also, this won't be nice with statics - more than one proxy -> they will _all_ have wonky retry numbers (write a test where you have two proxies to convince yourself that this won't work with a static).

**Review by Jolyon-S — COMMENTED** [review]
Static looks gnarly.

File: atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxy.java (hunk 7 above)
Comment (anjor) [reply atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxy.java:35]: oh you are right. This shouldn't be static.

**Pull request merged**